jobexec.1
.\" Manpage for jobexec.
.\" Contact adaulab at gmail dot com to correct errors or typos.
.TH jobexec 1 "13 Jun 2014" "0.1"
.SH NAME
jobexec \- command line batch job executor
.SH SYNOPSYS
jobexec <command> <cmd param 1> <cmd param 2> ...
.br
jobexec "<command with parameters to be executed through a shell>"

.SH DESCRIPTION
The aim of this utility is to be invoqued from a cron or from other unattended process as a way to keep track of the execution of a batch job. It will execute its parameters and redirect stdout and 
stderr to files in a directory created into /var/tmp/jobexec/

The idea is for this directory to contain the running outputs of batch jobs, so their execution can be audited and or monitored.

.SH FILES
.TP
.B /var/tmp/jobexec/<processname>.<hash>/<timestamp>/ 
"basedir" directory where the following status files related to the job will be left
.TP
.B <basedir>/command
the command line that was run
.TP
.B <basedir>/stdout
stdout job output
.TP
.B <basedir>/stderr
stderr job output
.TP
.B <basedir>/running
if present, the job is running, and the file contains the pid
.TP
.B <basedir>/finished
if present, the job finished, and the file contains the error level
.TP
.B <basedir>/exec_time
time taken by the job to finish execution
.TP
.B <basedir>/failed
if present, the job failed to run (it couldnt be executed)
.TP
.B <basedir>/bad_exit_status
if present, the job finished with an exit status other than 0

.SH AUTHOR
Alejandro Dau \- adaulab at gmail dot com

.SH BUGS
It will leave output in disk, so it will accumulate over time. You can leave the expiration of the files up
to the OS (some Unixes have a regular process that deletes unaccesed files in the /tmp and /var/tmp dirs). 
Or you can add the following to your crontab to delete all reports older than 5 days:

.B 7 5 * * *  jobexec find /var/tmp/jobexec/ -mtime +5 -delete -print

jobexec does not currently support Unicode parameters. As a workaround you can put them in a shell script and inboke that script from jobexec. 
 



jobexec.c
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <time.h>
#include "timer.h"
#include "jobexec.h"


// JIDSIZE is the size of the string to be used as job id (directory)
#define JIDSIZE 16


extern char **environ;
char dir[MAXDIRLEN];


unsigned int getParamsLen(char **params)
{
    unsigned int c;
    int j, i;

    c = 0;
    j = 0;
    while (params[j] != NULL) {
	c += 2;			// quotation marks
	c++;			// space
	i = 0;
	while (params[j][i] != '\0') {
	    c++;
	    i++;
	}
	j++;
    }
    return c;
}

char **getShellParams(char **params)
{
    static char *newparams[3];
    char *cmd;

    newparams[0] = SHELL;
    newparams[1] = "-c";

    int maxlen = getParamsLen(params);
    maxlen += 1;

    cmd = malloc(maxlen * sizeof(char));
    assertnlog2(cmd != NULL, "Failed to malloc for shell execution of",
		params[1]);

    int i, j, l;

    i = 0;
    j = 0;
    while (params[j] != NULL && i < maxlen) {
	l = 0;
	// if its a parameter to the command, use quotation marks
	if (j != 0 && i < maxlen)
	    cmd[i++] = '"';
	while (params[j][l] != '\0' && i < maxlen) {
	    if (params[j][l] == '"')
		cmd[i++] = '\\';
	    if (i < maxlen)
		cmd[i++] = params[j][l];
	    l++;
	}
	// close quotation marks
	if (j != 0 && i < maxlen)
	    cmd[i++] = '"';
	if (i < maxlen)
	    cmd[i++] = ' ';

	j++;
    }
    assertnlog(i < maxlen, "cmd maxlen calculation error!");

    cmd[i] = '\0';
    newparams[2] = cmd;

    return newparams;
}

void createOutputDir(int argc, char *argv[])
{

    int l, i, j;
    char cmdline[MAXCMDHASHLEN];
    char c;
    char *hash;

    j = 0;
    for (l = 1; l < argc; l++) {
	i = 0;
	while (argv[l][i] != '\0' && j < (MAXCMDHASHLEN - 1)) {
	    c = argv[l][i];
	    if (c >= 'A' && c <= 'Z'
		|| c >= 'a' && c <= 'z' || c >= '0' && c <= '9') {
		cmdline[j] = c;
		j++;
	    }
	    // the next if is to skip the path of the command in the dirname
	    if (l == 1 && c == '/')
		j = 0;

	    i++;
	}
    }

    cmdline[j] = '\0';
    hash = hashstr(cmdline, 5);
    assertnlog(hash != NULL, "malloc hashstr");

    // create jobexec dir if it doesnt exist
    snprintf(dir, sizeof(dir), "%s", BASEDIR);
    if (access(dir, F_OK) == -1) {
	assertnlog2(!mkdir(dir, S_IRWXU), "mkdir BASEDIR", dir);
	assertnlog2(!chmod(dir, S_IRWXU | S_IRWXG | S_IRWXO | S_ISVTX),
		    "chmod BASEDIR", dir);
    }

    l = strlen(dir);
    //shorten the cmdline string to be used in dir
    //create jobid dir
    cmdline[JIDSIZE] = '\0';
    snprintf(&dir[l], sizeof(dir) - l, "/%s-%s", cmdline, hash);
    if (access(dir, F_OK) == -1) {
	assertnlog2(!mkdir(dir, 00755), "mkdir dir", dir);
    }
    // create timestamp dir
    l = strlen(dir);
    snprintf(&dir[l], sizeof(dir) - l, "/%ju", time(NULL));
    if (access(dir, F_OK) == -1) {
	// create dir
	assertnlog2(!mkdir(dir, 00755), "mkdir dir", dir);
    }


}


int main(int argc, char *argv[])
{

    int fd1, fd2, l;
    FILE *f;
    char tmpstr[sizeof(dir) + 20];

    // usage:
    if (argc < 2) {
	printf("usage: %s <command> <cmd param 1> <cmd param 2> ...\n",
	       argv[0]);
	printf("\nmore information: man %s\n", argv[0]);
	exit(0);
    }

    createOutputDir(argc, argv);

    // write command file
    snprintf(tmpstr, sizeof(tmpstr), "%s/command", dir);
    f = fopen(tmpstr, "w");
    assertnlog2(f != NULL, "fopen outfile", tmpstr);

    for (l = 1; l < argc; l++) {
	fprintf(f, "%s ", argv[l]);
    }
    fprintf(f, "\n");
    fclose(f);


// create stdout file
    snprintf(tmpstr, sizeof(tmpstr), "%s/stdout", dir);
    fd1 = open(tmpstr, O_WRONLY | O_CREAT, 00644);
    assertnlog2(fd1 >= 0, "fopen outfile", tmpstr);

// create stderr file
    snprintf(tmpstr, sizeof(tmpstr), "%s/stderr", dir);
    fd2 = open(tmpstr, O_WRONLY | O_CREAT, 00644);
    assertnlog2(fd2 >= 0, "fopen outfile", tmpstr);


    pid_t childPID;
    childPID = fork();

    assertnlog(childPID >= 0, "fork");

    if (childPID == 0)		// child process
    {
	assertnlog(dup2(fd1, STDOUT_FILENO) != -1, "error duping stdout");
	assertnlog(dup2(fd2, STDERR_FILENO) != -1, "error duping stderr");

	close(fd1);
	close(fd2);
	char *cmd;
	char **params;

	cmd = argv[1];
	params = &argv[1];

	// if there is no path specified for command, invoke trhough shell
	if (cmd[0] != '/') {
	    cmd = SHELL;
	    params = getShellParams(params);
	}

	execve(cmd, params, environ);
	assertnlog2(0, "Exec failed, command not found", argv[1]);
	// shouldnt be reachable, but just in case:
	exit(ERRJOBEXEC);
    } else			//Parent process
    {
	// create 'running' file
	snprintf(tmpstr, sizeof(tmpstr), "%s/running", dir);
	f = fopen(tmpstr, "w");
	assertnlog2(f != NULL, "fopen failed", tmpstr);
	fprintf(f, "%d\n", childPID);
	fclose(f);

	struct timeval timer;
	timerStart(&timer);

#ifdef SYSLOG_ON
	syslog(LOG_INFO, "starting %s... Output: %s", argv[1], dir);
#endif

	int status;
	wait(&status);

	double exectime = timerGet(&timer);
	// delete 'running' file
	unlink(tmpstr);

	// save 'exec_time'
	snprintf(tmpstr, sizeof(tmpstr), "%s/exec_time", dir);
	f = fopen(tmpstr, "w");
	assertnlog2(f != NULL, "fopen failed", tmpstr);
	fprintf(f, "%.6f\n", exectime);
	fclose(f);


	// if exit status is ERRJOBEXEC, command failed to execute
	if (WEXITSTATUS(status) == ERRJOBEXEC) {

	    // create 'failed' file
	    snprintf(tmpstr, sizeof(tmpstr), "%s/failed", dir);
	    f = fopen(tmpstr, "w");
	    assertnlog2(f != NULL, "fopen failed", tmpstr);
	    fclose(f);

#ifdef SYSLOG_ON
	    syslog(LOG_ERR, "%s failed to execute. Output: %s", argv[1],
		   dir);
#endif

	    exit(-1);
	}
//          printf("\n cmd %s pid %d exited %d with status %d (%d)\n", argv[1], childPID, WIFEXITED(status), WEXITSTATUS(status), status);
	// write finished file
	snprintf(tmpstr, sizeof(tmpstr), "%s/finished", dir);
	f = fopen(tmpstr, "w");
	assertnlog2(f != NULL, "fopen finished", tmpstr);
	fprintf(f, "%d\n", WEXITSTATUS(status));
	fclose(f);

	// write bad_exit_status if status != 0
	if (WEXITSTATUS(status) != 0) {
	    snprintf(tmpstr, sizeof(tmpstr), "%s/bad_exit_status", dir);
	    f = fopen(tmpstr, "w");
	    assertnlog2(f != NULL, "fopen ", tmpstr);
	    fprintf(f, "%d\n", WEXITSTATUS(status));
	    fclose(f);
	}

    }

    return 0;
}
